service/collector/build.go

```diff
@@ -3,40 +3,121 @@ package collector
 import (
 	"github.com/giantswarm/microerror"
 	"github.com/giantswarm/micrologger"
+	"github.com/jszwedko/go-circleci"
 	"github.com/prometheus/client_golang/prometheus"
 )
 
+var (
+	buildLabelDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystem, "build"),
+		"circle CI build labels.",
+		[]string{
+			labelBranch,
+			labelRepo,
+			labelState,
+		},
+		nil,
+	)
+)
+
+var (
+	buildDesc = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    prometheus.BuildFQName(namespace, subsystem, "buildtime"),
+			Help:    "circle CI build time duration.",
+			Buckets: prometheus.ExponentialBuckets(bucketStart, bucketFactor, numBuckets),
+		},
+		[]string{
+			labelBranch,
+			labelRepo,
+			labelState,
+		},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(buildDesc)
+}
+
+type key struct {
+	Branch string
+	Repo   string
+	Status string
+}
+
 // BuildConfig is this collector's configuration struct.
 type BuildConfig struct {
-	Logger micrologger.Logger
+	CircleCIClient *circleci.Client
+	Logger         micrologger.Logger
 }
 
 // Build is the main struct for this collector.
 type Build struct {
-	logger micrologger.Logger
+	circleCIClient *circleci.Client
+	logger         micrologger.Logger
 }
 
 // NewAppResource creates a new Build metrics collector
 func NewBuild(config BuildConfig) (*Build, error) {
+	if config.CircleCIClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CircleCIClient must not be empty", config)
+	}
 	if config.Logger == nil {
 		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
 	}
 
 	c := &Build{
-		logger: config.Logger,
+		circleCIClient: config.CircleCIClient,
+		logger:         config.Logger,
 	}
 
 	return c, nil
 }
 
 // Collect is the main metrics collection function.
 func (c *Build) Collect(ch chan<- prometheus.Metric) error {
-	// TODO
+	builds, err := c.circleCIClient.ListRecentBuilds(100, 0)
```

Comment:
What about parameterizing these arguments as constants? It easier to tune later

Reply:
Actually, this 100 is their maximum number we can request, but parameterizing makes sense. 😆